week-2/mandatory/3-api/server.js

```diff
@@ -0,0 +1,382 @@
+      res.status(500).json(serverError);
+      throw error;
+    } else {
+      res.status(200).json(result.rows);
+    }
+  });
+});
+
+app.get("/customers/:customerID", (req, res) => {
+  let id = parseInt(req.params.customerID);
+  db.query("SELECT * FROM customers where id = $1", [id], (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+      throw error;
+    } else {
+      if (result.rows.length > 0) {
+        res.status(200).json(result.rows[0]);
+      } else {
+        res.status(404).json(`There is no customer with this ID ${id}`);
+      }
+    }
+  });
+});
+
+app.get("/customers/by-name/:name", (req, res) => {
+  let name = req.params.name.toLowerCase();
+
+  db.query(
+    "SELECT * FROM customers where Lower(name) LIKE $1 || '%'",
+    [name],
+    (error, result) => {
+      if (error) {
+        console.log("pg error code:", error.code);
+        res.status(500).json(serverError);
+        throw error;
+      } else {
+        res.status(200).json(result.rows);
+      }
+    }
+  );
+});
+
+app.get("/suppliers", (req, res) => {
+  db.query("SELECT * FROM suppliers", (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+    } else {
+      res.status(200).json(result.rows);
+    }
+  });
+});
+
+app.get("/products", (req, res) => {
+  let name = req.query.name;
+  if (name) {
+    name.toLowerCase();
+    db.query(
+      "SELECT p.product_name, a.unit_price , s.supplier_name from products p join product_availability a on p.id = a.prod_id join suppliers s on s.id = a.supp_id where Lower(p.product_name) like '%'||$1||'%'",
+      [name],
+      (error, result) => {
+        if (error) {
+          console.log("pg error code:", error.code);
+          res.status(500).json(serverError);
+        } else {
+          res.status(200).json(result.rows);
+        }
+      }
+    );
+  } else {
+    db.query(
+      "SELECT p.product_name, a.unit_price , s.supplier_name from products p join product_availability a on p.id = a.prod_id join suppliers s on s.id = a.supp_id",
+      (error, result) => {
+        if (error) {
+          console.log("pg error code:", error.code);
+          res.status(500).json(serverError);
+        } else {
+          res.status(200).json(result.rows);
+        }
+      }
+    );
+  }
+});
+
+app.post("/customers", (req, res) => {
+  let name = req.body.name,
+    address = req.body.address,
+    city = req.body.city,
+    country = req.body.country;
+  let sql =
+    "INSERT INTO customers (name, address, city, country) " +
+    "VALUES ($1, $2, $3, $4) returning id";
+  let attributes = [name, address, city, country];
+
+  db.query(sql, attributes, (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+    } else {
+      res
+        .status(201)
```

Comment:
Good use of this HTTP status - 200 would have been fine, but 201 is more specific.